README.md
# mikunyan

Library to deserialize Unity AssetBundle files (\*.unity3d) and asset files.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'mikunyan'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install mikunyan

If you want to install development build:

    $ git clone https://github.com/Ishotihadus/mikunyan
    $ cd mikunyan
    $ rake build
    $ gem install -l pkg/mikunyan-3.9.x.gem

## Usage

### Basic Usage

```ruby
require 'mikunyan'

# load AssetBundle
bundle = Mikunyan::AssetBundle.file(filename)

# you can load AssetBundle from blob
# bundle = Mikunyan::AssetBundle.load(blob)

# select asset (normaly only one asset)
asset = bundle.assets[0]

# or you can directly load asset
# asset = Mikunyan::Asset.file(filename)

# object list
list = asset.objects

# object PathIds
path_ids = asset.path_ids

# object container table (if available)
containers = asset.containers

# load object (Mikunyan::ObjectValue)
obj = asset.parse_object(path_ids[0])

# simplified structure (based on Hash)
obj_hash = asset.parse_object_simple(path_ids[0])

# hash can be easily serialized to json
require 'json'
obj_hash.to_json
```

### Mikunyan::ObjectValue

`Mikunyan::ObjectValue` can be 3 types. Value, array and map.

```ruby
# You can get whether obj is value or not
obj.value?

# get value
obj.value

# same as obj.value
obj[]


# You can get whether obj is array or not
obj.array?

# get array
obj.value

# you can directly access by index
obj[0]


# If obj is map, you can get keys
obj.keys

# get child object
obj[key]

# same as obj[key]
obj.key
```

### Unpack Texture2D

You can get png file directly from Texture2D asset. Output object's class is `ChunkyPNG::Image`.

Acceptable format is basic texture formats (1, 2, 3, 4, 5, 7 and 13) and ETC_RGB4 (34).

```ruby
require 'mikunyan/decoders'

# get some Texture2D asset
obj = asset.parse_object(path_ids[1])

# you can get Image object
img = Mikunyan::ImageDecoder.decode_object(obj)

# save it!
img.save('mikunyan.png')
```

### Json / YAML Outputer

`mikunyan-json` is an executable command for converting unity3d to json.

    $ mikunyan-json bundle.unity3d > bundle.json

Available options:

- `--as-asset` (`-a`): interpret input file as not AssetBudnle but Asset
- `--pretty` (`-p`): prettify output json (`mikunyan-json` only)
- `--yaml` (`-y`): YAML mode

## Dependencies

- [json](https://rubygems.org/gems/json)
- [extlz4](https://rubygems.org/gems/extlz4)
- [bin_utils](https://rubygems.org/gems/bin_utils)
- [chunky_png](https://rubygems.org/gems/chunky_png)

Mikunyan uses [oily_png](https://rubygems.org/gems/oily_png) instead of chunky_png if available.

## FAQ

### Sometimes unpacking fails

I'm sorry...

### Can I unpack Mesh files?

It's hard work for me...

### What mikunyan comes from?

[Miku Maekawa](http://www.project-imas.com/wiki/Miku_Maekawa).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/Ishotihadus/mikunyan.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Change log of mikunyan

## 3.9.0 - 2017-07-05

- First release


lib/mikunyan/constants.rb
module Mikunyan
    STRING_TABLE = {
        0=>'AABB',
        5=>'AnimationClip',
        19=>'AnimationCurve',
        34=>'AnimationState',
        49=>'Array',
        55=>'Base',
        60=>'BitField',
        69=>'bitset',
        76=>'bool',
        81=>'char',
        86=>'ColorRGBA',
        96=>'Component',
        106=>'data',
        111=>'deque',
        117=>'double',
        124=>'dynamic_array',
        138=>'FastPropertyName',
        155=>'first',
        161=>'float',
        167=>'Font',
        172=>'GameObject',
        183=>'Generic Mono',
        196=>'GradientNEW',
        208=>'GUID',
        213=>'GUIStyle',
        222=>'int',
        226=>'list',
        231=>'long long',
        241=>'map',
        245=>'Matrix4x4f',
        256=>'MdFour',
        263=>'MonoBehaviour',
        277=>'MonoScript',
        288=>'m_ByteSize',
        299=>'m_Curve',
        307=>'m_EditorClassIdentifier',
        331=>'m_EditorHideFlags',
        349=>'m_Enabled',
        359=>'m_ExtensionPtr',
        374=>'m_GameObject',
        387=>'m_Index',
        395=>'m_IsArray',
        405=>'m_IsStatic',
        416=>'m_MetaFlag',
        427=>'m_Name',
        434=>'m_ObjectHideFlags',
        452=>'m_PrefabInternal',
        469=>'m_PrefabParentObject',
        490=>'m_Script',
        499=>'m_StaticEditorFlags',
        519=>'m_Type',
        526=>'m_Version',
        536=>'Object',
        543=>'pair',
        548=>'PPtr<Component>',
        564=>'PPtr<GameObject>',
        581=>'PPtr<Material>',
        596=>'PPtr<MonoBehaviour>',
        616=>'PPtr<MonoScript>',
        633=>'PPtr<Object>',
        646=>'PPtr<Prefab>',
        659=>'PPtr<Sprite>',
        672=>'PPtr<TextAsset>',
        688=>'PPtr<Texture>',
        702=>'PPtr<Texture2D>',
        718=>'PPtr<Transform>',
        734=>'Prefab',
        741=>'Quaternionf',
        753=>'Rectf',
        759=>'RectInt',
        767=>'RectOffset',
        778=>'second',
        785=>'set',
        789=>'short',
        795=>'size',
        800=>'SInt16',
        807=>'SInt32',
        814=>'SInt64',
        821=>'SInt8',
        827=>'staticvector',
        840=>'string',
        847=>'TextAsset',
        857=>'TextMesh',
        866=>'Texture',
        874=>'Texture2D',
        884=>'Transform',
        894=>'TypelessData',
        907=>'UInt16',
        914=>'UInt32',
        921=>'UInt64',
        928=>'UInt8',
        934=>'unsigned int',
        947=>'unsigned long long',
        966=>'unsigned short',
        981=>'vector',
        988=>'Vector2f',
        997=>'Vector3f',
        1006=>'Vector4f',
        1015=>'m_ScriptingClassIdentifier',
        1042=>'Gradient'
    }

    CLASS_ID = {
        1=>'GameObject',
        2=>'Component',
        3=>'LevelGameManager',
        4=>'Transform',
        5=>'TimeManager',
        6=>'GlobalGameManager',
        8=>'Behaviour',
        9=>'GameManager',
        11=>'AudioManager',
        12=>'ParticleAnimator',
        13=>'InputManager',
        15=>'EllipsoidParticleEmitter',
        17=>'Pipeline',
        18=>'EditorExtension',
        19=>'Physics2DSettings',
        20=>'Camera',
        21=>'Material',
        23=>'MeshRenderer',
        25=>'Renderer',
        26=>'ParticleRenderer',
        27=>'Texture',
        28=>'Texture2D',
        29=>'SceneSettings',
        30=>'GraphicsSettings',
        33=>'MeshFilter',
        41=>'OcclusionPortal',
        43=>'Mesh',
        45=>'Skybox',
        47=>'QualitySettings',
        48=>'Shader',
        49=>'TextAsset',
        50=>'Rigidbody2D',
        51=>'Physics2DManager',
        53=>'Collider2D',
        54=>'Rigidbody',
        55=>'PhysicsManager',
        56=>'Collider',
        57=>'Joint',
        58=>'CircleCollider2D',
        59=>'HingeJoint',
        60=>'PolygonCollider2D',
        61=>'BoxCollider2D',
        62=>'PhysicsMaterial2D',
        64=>'MeshCollider',
        65=>'BoxCollider',
        66=>'SpriteCollider2D',
        68=>'EdgeCollider2D',
        72=>'ComputeShader',
        74=>'AnimationClip',
        75=>'ConstantForce',
        76=>'WorldParticleCollider',
        78=>'TagManager',
        81=>'AudioListener',
        82=>'AudioSource',
        83=>'AudioClip',
        84=>'RenderTexture',
        87=>'MeshParticleEmitter',
        88=>'ParticleEmitter',
        89=>'Cubemap',
        90=>'Avatar',
        91=>'AnimatorController',
        92=>'GUILayer',
        93=>'RuntimeAnimatorController',
        94=>'ScriptMapper',
        95=>'Animator',
        96=>'TrailRenderer',
        98=>'DelayedCallManager',
        102=>'TextMesh',
        104=>'RenderSettings',
        108=>'Light',
        109=>'CGProgram',
        110=>'BaseAnimationTrack',
        111=>'Animation',
        114=>'MonoBehaviour',
        115=>'MonoScript',
        116=>'MonoManager',
        117=>'Texture3D',
        118=>'NewAnimationTrack',
        119=>'Projector',
        120=>'LineRenderer',
        121=>'Flare',
        122=>'Halo',
        123=>'LensFlare',
        124=>'FlareLayer',
        125=>'HaloLayer',
        126=>'NavMeshAreas',
        127=>'HaloManager',
        128=>'Font',
        129=>'PlayerSettings',
        130=>'NamedObject',
        131=>'GUITexture',
        132=>'GUIText',
        133=>'GUIElement',
        134=>'PhysicMaterial',
        135=>'SphereCollider',
        136=>'CapsuleCollider',
        137=>'SkinnedMeshRenderer',
        138=>'FixedJoint',
        140=>'RaycastCollider',
        141=>'BuildSettings',
        142=>'AssetBundle',
        143=>'CharacterController',
        144=>'CharacterJoint',
        145=>'SpringJoint',
        146=>'WheelCollider',
        147=>'ResourceManager',
        148=>'NetworkView',
        149=>'NetworkManager',
        150=>'PreloadData',
        152=>'MovieTexture',
        153=>'ConfigurableJoint',
        154=>'TerrainCollider',
        155=>'MasterServerInterface',
        156=>'TerrainData',
        157=>'LightmapSettings',
        158=>'WebCamTexture',
        159=>'EditorSettings',
        160=>'InteractiveCloth',
        161=>'ClothRenderer',
        162=>'EditorUserSettings',
        163=>'SkinnedCloth',
        164=>'AudioReverbFilter',
        165=>'AudioHighPassFilter',
        166=>'AudioChorusFilter',
        167=>'AudioReverbZone',
        168=>'AudioEchoFilter',
        169=>'AudioLowPassFilter',
        170=>'AudioDistortionFilter',
        171=>'SparseTexture',
        180=>'AudioBehaviour',
        181=>'AudioFilter',
        182=>'WindZone',
        183=>'Cloth',
        184=>'SubstanceArchive',
        185=>'ProceduralMaterial',
        186=>'ProceduralTexture',
        191=>'OffMeshLink',
        192=>'OcclusionArea',
        193=>'Tree',
        194=>'NavMeshObsolete',
        195=>'NavMeshAgent',
        196=>'NavMeshSettings',
        197=>'LightProbesLegacy',
        198=>'ParticleSystem',
        199=>'ParticleSystemRenderer',
        200=>'ShaderVariantCollection',
        205=>'LODGroup',
        206=>'BlendTree',
        207=>'Motion',
        208=>'NavMeshObstacle',
        210=>'TerrainInstance',
        212=>'SpriteRenderer',
        213=>'Sprite',
        214=>'CachedSpriteAtlas',
        215=>'ReflectionProbe',
        216=>'ReflectionProbes',
        218=>'Terrain',
        220=>'LightProbeGroup',
        221=>'AnimatorOverrideController',
        222=>'CanvasRenderer',
        223=>'Canvas',
        224=>'RectTransform',
        225=>'CanvasGroup',
        226=>'BillboardAsset',
        227=>'BillboardRenderer',
        228=>'SpeedTreeWindAsset',
        229=>'AnchoredJoint2D',
        230=>'Joint2D',
        231=>'SpringJoint2D',
        232=>'DistanceJoint2D',
        233=>'HingeJoint2D',
        234=>'SliderJoint2D',
        235=>'WheelJoint2D',
        238=>'NavMeshData',
        240=>'AudioMixer',
        241=>'AudioMixerController',
        243=>'AudioMixerGroupController',
        244=>'AudioMixerEffectController',
        245=>'AudioMixerSnapshotController',
        246=>'PhysicsUpdateBehaviour2D',
        247=>'ConstantForce2D',
        248=>'Effector2D',
        249=>'AreaEffector2D',
        250=>'PointEffector2D',
        251=>'PlatformEffector2D',
        252=>'SurfaceEffector2D',
        258=>'LightProbes',
        271=>'SampleClip',
        272=>'AudioMixerSnapshot',
        273=>'AudioMixerGroup',
        290=>'AssetBundleManifest',
        1001=>'Prefab',
        1002=>'EditorExtensionImpl',
        1003=>'AssetImporter',
        1004=>'AssetDatabase',
        1005=>'Mesh3DSImporter',
        1006=>'TextureImporter',
        1007=>'ShaderImporter',
        1008=>'ComputeShaderImporter',
        1011=>'AvatarMask',
        1020=>'AudioImporter',
        1026=>'HierarchyState',
        1027=>'GUIDSerializer',
        1028=>'AssetMetaData',
        1029=>'DefaultAsset',
        1030=>'DefaultImporter',
        1031=>'TextScriptImporter',
        1032=>'SceneAsset',
        1034=>'NativeFormatImporter',
        1035=>'MonoImporter',
        1037=>'AssetServerCache',
        1038=>'LibraryAssetImporter',
        1040=>'ModelImporter',
        1041=>'FBXImporter',
        1042=>'TrueTypeFontImporter',
        1044=>'MovieImporter',
        1045=>'EditorBuildSettings',
        1046=>'DDSImporter',
        1048=>'InspectorExpandedState',
        1049=>'AnnotationManager',
        1050=>'PluginImporter',
        1051=>'EditorUserBuildSettings',
        1052=>'PVRImporter',
        1053=>'ASTCImporter',
        1054=>'KTXImporter',
        1101=>'AnimatorStateTransition',
        1102=>'AnimatorState',
        1105=>'HumanTemplate',
        1107=>'AnimatorStateMachine',
        1108=>'PreviewAssetType',
        1109=>'AnimatorTransition',
        1110=>'SpeedTreeImporter',
        1111=>'AnimatorTransitionBase',
        1112=>'SubstanceImporter',
        1113=>'LightmapParameters',
        1120=>'LightmapSnapshot'
    }
end
